cloud-controller-manager/do/firewall_controller.go

```diff
@@ -182,12 +194,37 @@ func (fm *firewallManager) Get(ctx context.Context) (*godo.Firewall, error) {
 	// check cache and query the API firewall service to get firewall by ID, if it exists. Return it. If not, continue.
 	fw := fm.fwCache.getCachedFirewall()
 	if fw != nil {
-		fw, resp, err := fm.client.Firewalls.Get(ctx, fw.ID)
+		var (
+			resp *godo.Response
+			err  error
+		)
+		fw, resp, err := func() (*godo.Firewall, *godo.Response, error) {
+			var (
+				code   int
+				method string
+			)
+			// The ObserverFunc gets called by the deferred ObserveDuration. The
+			// method and code values will be set before ObserveDuration is called
+			// with the value returned from the response from the Firewall API request.
+			timer := prometheus.NewTimer(prometheus.ObserverFunc(func(v float64) {
+				fm.metrics.apiRequestDuration.With(prometheus.Labels{"method": method, "code": strconv.FormatInt(int64(code), 10)}).Observe(v)
+			}))
+			defer timer.ObserveDuration()
+			fw, resp, err := fm.client.Firewalls.Get(ctx, fw.ID)
+			if resp != nil {
+				code = resp.StatusCode
```

Comment:
Whoops, I just about overlooked the fact that here we are also reassigning a variable whose value has already been captured by the intended `Observer` (refer the explanation I gave in my previous review).

To fix this, instead of creating `code` and `method` variables I would do the same thing that is being done in `observeRunLoopDuration` and use a `prometheus.Labels` variable whose value you would update here (ie `labels["code"] = resp.StatusCode`). This comment applies everywhere the same timer/observer pattern is being applied, ie the changes starting at lines 239, 287, 372, and 398.

Reply:
Actually, forget my previous comment I see now that you're using `prometheus.ObserverFunc` with a callback that will capture the values of `code` and `method` after they have been set and before the deferred `timer.ObserveDuration` has run.